okhttp/src/main/java/com/backblaze/b2/client/okHttpClient/B2OkHttpClientImpl.java

```diff
@@ -0,0 +1,327 @@
+            if (!response.isSuccessful()) {
+                throw extractExceptionFromErrorResponse(response, response.body().string());
+            } else {
+                if( response.code() == 200){
+                    Headers responseHeaders = response.headers();
+                    return makeB2Headers(responseHeaders);
+                } else {
+                    throw B2Exception.create(null, response.code(), null, "");
+                }
+            }
+        } catch (IOException e) {
+            throw translateToB2Exception(e, url);
+        }
+    }
+
+    @Override
+    public void close() {
+        // Closing response body streams is handled elsewhere by OkHttp.
+    }
+
+    private String postJsonAndReturnString(String url,
+                                           B2Headers headersOrNull,
+                                           Object request) throws B2Exception {
+        try {
+            B2Json bzJson = B2Json.get();
+            String requestJson = bzJson.toJson(request);
+            byte[] requestBytes = getUtf8Bytes(requestJson);
+            return postAndReturnString(url, headersOrNull,new ByteArrayInputStream(requestBytes), requestJson.length());
+        } catch (B2JsonException e) {
+            //log.warn("Unable to serialize " + request.getClass() + " using B2Json, was passed in request for " + url, ex);
+            throw new B2LocalException("parsing_failed", "B2Json.toJson(" + request.getClass() + ") failed: " + e.getMessage(), e);
+        }
+
+    }
+
+    private byte[] readFully(InputStream inputStream, int contentLength) throws IOException{
+        int numRead = 0;
+        int offset = 0;
+        int length = contentLength;
+        byte[] bytes = new byte[contentLength];
+        while( numRead < contentLength ){
+            int n = inputStream.read(bytes, offset, length);
+            if( n > -1 ){
+                numRead += n;
+                offset += n;
+                length = contentLength - numRead;
+            }
+        }
+        return bytes;
+    }
+
+    /**
+     * POSTs to a web service that returns content, and returns the content
+     * as a single string.
+     *
+     * @param url the url to post to
+     * @param headersOrNull the headers, if any.
+     *
+     * @return the body of the response.
+     * @throws B2Exception if there's any trouble
+     */
+    private  String postAndReturnString( String url,  B2Headers headersOrNull,  InputStream inputStream, long contentLength)
+            throws B2Exception {
+        RequestBody body;
+        try {
+            if( contentLength <= (Integer.MAX_VALUE-HOTSPOT_FUDGE) ){
+                byte[] bytes = readFully(inputStream, (int) contentLength);
+                body = RequestBody.create(MediaType.get(APPLICATION_JSON), bytes);
+            } else {
+                String tempFileName = UUID.randomUUID().toString();
+                File tempFile = File.createTempFile(tempFileName, ".tmp");
```

Comment:
i think we should call tempFile.deleteOnExit() immediately after this line so that we don't leak it if we fail part of the way through the copy() below.

Reply:
it also seems like we should hold on to this temp file name and delete it in a finally block that covers the whole method.  we don't really want to wait until the app exits to delete the file.

Reply:
the documentation[1] says that Files.createTempFile() "may be more suited to security-sensitive applications".  (note that's on Files, not File).  should we be using that instead?

[1] https://developer.android.com/reference/java/io/File#createTempFile(java.lang.String,%2520java.lang.String,%2520java.io.File)

Reply:
hey @OldSkoolMark --

you addressed my second and third comment in this area, but i'm not sure you saw my first one...

Reply:
I refactored this part of the code a bit. I think your concerns are now addressed.